Mathlib/Tactic/ProdAssoc.lean

```diff
@@ -0,0 +1,150 @@
+/-
+Copyright (c) 2023 Adam Topaz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Adam Topaz
+-/
+import Mathlib.Lean.Expr.Basic
+import Mathlib.Logic.Equiv.Basic
+
+/-!
+# Associativity of products
+
+This file constructs a term elaborator for "obvious" equivalences between iterated products.
+For example,
+```lean
+(prod_assoc% : (α × β) × (γ × δ) ≃ α × (β × γ) × δ)
+```
+gives the "obvious" equivalence between `(α × β) × (γ × δ)` and `α × (β × γ) × δ`.
+-/
+
+namespace Lean.Expr
+
+open Lean Meta
+
+/-- A helper type to keep track of universe levels and types in iterated produts. -/
+inductive ProdTree where
+  | type (tp : Expr) (l : Level)
+  | prod (fst snd : ProdTree) (lfst lsnd : Level)
+deriving Repr
+
+/-- The iterated product corresponding to a `ProdTree`. -/
+def ProdTree.getType : ProdTree → Expr
+  | type tp _ => tp
+  | prod fst snd u v => mkAppN (.const ``Prod [u,v]) #[fst.getType, snd.getType]
+
+/-- The number of types appearing in an iterated product encoded as a `ProdTree`. -/
+def ProdTree.size : ProdTree → Nat
+  | type _ _ => 1
+  | prod fst snd _ _ => fst.size + snd.size
+
+/-- The components of an interated product, presented as a `ProdTree`. -/
+def ProdTree.components : ProdTree → List Expr
+  | type tp _ => [tp]
+  | prod fst snd _ _ => fst.components ++ snd.components
+
+/-- Make a `ProdTree` out of an `Expr`. -/
+def mkProdTree (e : Expr) : MetaM ProdTree :=
+  match e with
```

Comment:
It's a little overkill (`e.consumeMData` is probably all you need in practice), but if you want to avoid some metadata-related issues you probably should do
```suggestion
  match e.cleanupAnnotations with
```
You might even consider `whnfR` instead of `cleanupAnnotations` if you want this elaborator to unfold reducible definitions, but I'm assuming we don't want that, since it could be rather surprising.

Reply:
I did indeed purposefully avoid reduction. But I'll added `cleanupAnnotations` as suggested.